// @keystonehq/.eth-keyring-FTlocxIN/dist/eth-keyring.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var bcUrRegistryEth = require('@keystonehq/bc-ur-registry-eth');
var sdk = require('@keystonehq/sdk');
var sdk__default = _interopDefault(sdk);
var baseEthKeyring = require('@keystonehq/base-eth-keyring');

class DefaultInteractionProvider {
  constructor() {
    this.keystoneSDK = undefined;

    this.readCryptoHDKey = async () => {
      const decodedResult = await this.keystoneSDK.read([sdk.SupportedResult.UR_CRYPTO_HDKEY], {
        title: 'Sync Keystone',
        description: 'Please scan the QR code displayed on your Keystone',
        renderInitial: {
          walletMode: 'Web3',
          link: 'https://keyst.one/defi'
        },
        URTypeErrorMessage: 'The scanned QR code is not the sync code from the Keystone hardware wallet. Please verify the code and try again ( Keystone firmware V1.3.0 or newer required).'
      });

      if (decodedResult.status === sdk.ReadStatus.success) {
        const {
          result
        } = decodedResult;
        const cryptoHDKey = bcUrRegistryEth.CryptoHDKey.fromCBOR(result.cbor);
        return cryptoHDKey;
      } else {
        throw new Error('Reading canceled');
      }
    };

    this.requestSignature = async (ethSignRequest, requestTitle, requestDescription) => {
      const status = await this.keystoneSDK.play(ethSignRequest.toUR(), {
        hasNext: true,
        title: requestTitle,
        description: requestDescription
      });
      if (status === sdk.PlayStatus.canceled) throw new Error('#ktek_error[play-cancel]: play canceled');
      const result = await this.keystoneSDK.read([sdk.SupportedResult.UR_ETH_SIGNATURE], {
        title: 'Scan Keystone',
        description: 'Please scan the QR code displayed on your Keystone'
      });

      if (result.status === sdk.ReadStatus.canceled) {
        throw new Error('#ktek_error[read-cancel]: read signature canceled');
      } else {
        return bcUrRegistryEth.ETHSignature.fromCBOR(result.result.cbor);
      }
    };

    if (DefaultInteractionProvider.instance) {
      return DefaultInteractionProvider.instance;
    }

    sdk__default.bootstrap();
    this.keystoneSDK = sdk__default.getSdk();
    DefaultInteractionProvider.instance = this;
  }

}

class DefaultKeyring extends baseEthKeyring.BaseKeyring {
  constructor(opts) {
    super(opts);

    this.getInteraction = () => {
      return new DefaultInteractionProvider();
    };
  }

  static getEmptyKeyring() {
    return new DefaultKeyring({
      xfp: '',
      xpub: '',
      hdPath: '',
      perPage: 5,
      page: 0,
      accounts: [],
      currentAccount: 0,
      paths: {}
    });
  }

}
DefaultKeyring.type = baseEthKeyring.BaseKeyring.type;

exports.default = DefaultKeyring;
